lib/auth/auth_with_roles.go

```diff
@@ -2778,6 +2774,23 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 	return certs, nil
 }
 
+// verifyUserDeviceForCertIssuance verifies if the user device is a trusted
+// device, in accordance to the certificate usage and the cluster's DeviceTrust
+// settings. It's meant to be called before issuing new user certificates.
+// Each Node (or access server) verifies the device independently, so this check
+// is not paramount to the access system itself, but it stops bad attempts from
+// progressing further and provides better feedback than other protocol-specific
+// failures.
+func (a *ServerWithRoles) verifyUserDeviceForCertIssuance(usage proto.UserCertsRequest_CertUsage, dt *types.DeviceTrust) error {
```

Comment:
nit: maybe this fn can include:
```
	authPref, err := a.authServer.GetAuthPreference(ctx)
	if err != nil {
		return nil, trace.Wrap(err)
```

Reply:
Ack. I think I'd rather keep that outside, as callers could arguably use other info from the authPref themselves, plus we keep the parameters more targeted.